### PR TITLE
refactor: Make surface intersection tolerance configurable

### PR DESCRIPTION
**Pull request by andiwand**
`s_onSurfaceTolerance` is used in a bunch of places where the user might actually expect a configurable tolerance.

in this PR I add a tolerance parameter to `Surface::intersect` which can then be used by the navigator or target reached aborter

### CODE REVIEW

**Comment by asalzburger** [comment]
I like that! Doesn't that actually qualify as a breaking change?

Will do the review later on that.

**Comment by andiwand** [comment]
> I like that! Doesn't that actually qualify as a breaking change?

good point - I am not sure. if we are expecting the users to have their own surfaces, yes. if the only use the interface of the surface then the api should be stable because of the default value

**Review by asalzburger — CHANGES_REQUESTED** [review]
Can you change the `double` to `ActsScalar` please ?

File: Core/include/Acts/Surfaces/CylinderSurface.hpp [hunk 1]
```diff
@@ -177,14 +177,15 @@ class CylinderSurface : public Surface {
...
   ///
   /// @return SurfaceIntersection object (contains intersection & surface)
-  SurfaceIntersection intersect(const GeometryContext& gctx,
-                                const Vector3& position,
-                                const Vector3& direction,
-                                const BoundaryCheck& bcheck) const final;
+  SurfaceIntersection intersect(
+      const GeometryContext& gctx, const Vector3& position,
+      const Vector3& direction, const BoundaryCheck& bcheck = false,
+      double tolerance = s_onSurfaceTolerance) const final;
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/CylinderSurface.hpp:188]:
```suggestion
      ActsScalar tolerance = s_onSurfaceTolerance) const final;
```

File: Core/include/Acts/Surfaces/DiscSurface.hpp [hunk 2]
```diff
@@ -263,8 +264,8 @@ class DiscSurface : public Surface {
   /// @return The SurfaceIntersection object
   SurfaceIntersection intersect(
       const GeometryContext& gctx, const Vector3& position,
-      const Vector3& direction,
-      const BoundaryCheck& bcheck = false) const final;
+      const Vector3& direction, const BoundaryCheck& bcheck = false,
+      double tolerance = s_onSurfaceTolerance) const final;
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/DiscSurface.hpp:268]:
```suggestion
      ActsScalar tolerance = s_onSurfaceTolerance) const final;
```

File: Core/include/Acts/Surfaces/LineSurface.hpp [hunk 3]
```diff
@@ -228,8 +229,8 @@ class LineSurface : public Surface {
   /// @return is the intersection object
   SurfaceIntersection intersect(
       const GeometryContext& gctx, const Vector3& position,
-      const Vector3& direction,
-      const BoundaryCheck& bcheck = false) const final;
+      const Vector3& direction, const BoundaryCheck& bcheck = false,
+      double tolerance = s_onSurfaceTolerance) const final;
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/LineSurface.hpp:233]:
```suggestion
      ActsScalar tolerance = s_onSurfaceTolerance) const final;
```

File: Core/include/Acts/Surfaces/PlaneSurface.hpp [hunk 4]
```diff
@@ -177,8 +178,8 @@ class PlaneSurface : public Surface {
   /// @return the SurfaceIntersection object
   SurfaceIntersection intersect(
       const GeometryContext& gctx, const Vector3& position,
-      const Vector3& direction,
-      const BoundaryCheck& bcheck = false) const final;
+      const Vector3& direction, const BoundaryCheck& bcheck = false,
+      double tolerance = s_onSurfaceTolerance) const final;
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/PlaneSurface.hpp:182]:
```suggestion
      ActsScalar tolerance = s_onSurfaceTolerance) const final;
```

File: Core/include/Acts/Surfaces/Surface.hpp [hunk 5]
```diff
@@ -398,12 +398,13 @@ class Surface : public virtual GeometryObject,
...
   ///
   /// @return SurfaceIntersection object (contains intersection & surface)
-  virtual SurfaceIntersection intersect(const GeometryContext& gctx,
-                                        const Vector3& position,
-                                        const Vector3& direction,
-                                        const BoundaryCheck& bcheck) const = 0;
+  virtual SurfaceIntersection intersect(
+      const GeometryContext& gctx, const Vector3& position,
+      const Vector3& direction, const BoundaryCheck& bcheck = false,
+      double tolerance = s_onSurfaceTolerance) const = 0;
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/Surface.hpp:407]:
```suggestion
      ActsScalar tolerance = s_onSurfaceTolerance) const = 0;
```

File: Core/include/Acts/Surfaces/detail/PlanarHelper.hpp [hunk 6]
```diff
@@ -25,7 +25,7 @@ namespace PlanarHelper {
 /// @return The intersection
 inline Intersection3D intersect(const Transform3& transform,
                                 const Vector3& position,
-                                const Vector3& direction) {
+                                const Vector3& direction, double tolerance) {
```
Comment (andiwand) [inline Core/include/Acts/Surfaces/detail/PlanarHelper.hpp:28]:
```suggestion
                                const Vector3& direction, ActsScalar tolerance) {
```

File: Core/src/Surfaces/ConeSurface.cpp [hunk 7]
```diff
@@ -280,7 +280,8 @@ Acts::detail::RealQuadraticEquation Acts::ConeSurface::intersectionSolver(
 
 Acts::SurfaceIntersection Acts::ConeSurface::intersect(
     const GeometryContext& gctx, const Vector3& position,
-    const Vector3& direction, const BoundaryCheck& bcheck) const {
+    const Vector3& direction, const BoundaryCheck& bcheck,
+    double tolerance) const {
```
Comment (andiwand) [inline Core/src/Surfaces/ConeSurface.cpp:284]:
```suggestion
    ActsScalar tolerance) const {
```

File: Core/src/Surfaces/CylinderSurface.cpp [hunk 8]
```diff
@@ -210,7 +210,8 @@ Acts::detail::RealQuadraticEquation Acts::CylinderSurface::intersectionSolver(
 
 Acts::SurfaceIntersection Acts::CylinderSurface::intersect(
     const GeometryContext& gctx, const Vector3& position,
-    const Vector3& direction, const BoundaryCheck& bcheck) const {
+    const Vector3& direction, const BoundaryCheck& bcheck,
+    double tolerance) const {
```
Comment (andiwand) [inline Core/src/Surfaces/CylinderSurface.cpp:214]:
```suggestion
    ActsScalar tolerance) const {
```

File: Core/src/Surfaces/DiscSurface.cpp [hunk 9]
```diff
@@ -287,12 +287,13 @@ Acts::FreeToBoundMatrix Acts::DiscSurface::freeToBoundJacobian(
 
 Acts::SurfaceIntersection Acts::DiscSurface::intersect(
     const GeometryContext& gctx, const Vector3& position,
-    const Vector3& direction, const BoundaryCheck& bcheck) const {
+    const Vector3& direction, const BoundaryCheck& bcheck,
+    double tolerance) const {
```
Comment (andiwand) [inline Core/src/Surfaces/DiscSurface.cpp:291]:
```suggestion
    ActsScalar tolerance) const {
```

File: Core/src/Surfaces/LineSurface.cpp [hunk 10]
```diff
@@ -126,7 +126,8 @@ const Acts::SurfaceBounds& Acts::LineSurface::bounds() const {
 
 Acts::SurfaceIntersection Acts::LineSurface::intersect(
     const GeometryContext& gctx, const Vector3& position,
-    const Vector3& direction, const BoundaryCheck& bcheck) const {
+    const Vector3& direction, const BoundaryCheck& bcheck,
+    double tolerance) const {
```
Comment (andiwand) [inline Core/src/Surfaces/LineSurface.cpp:130]:
```suggestion
    ActsScalar tolerance) const {
```

File: Core/src/Surfaces/PlaneSurface.cpp [hunk 11]
```diff
@@ -172,12 +172,13 @@ double Acts::PlaneSurface::pathCorrection(const GeometryContext& gctx,
 
 Acts::SurfaceIntersection Acts::PlaneSurface::intersect(
     const GeometryContext& gctx, const Vector3& position,
-    const Vector3& direction, const BoundaryCheck& bcheck) const {
+    const Vector3& direction, const BoundaryCheck& bcheck,
+    double tolerance) const {
```
Comment (andiwand) [inline Core/src/Surfaces/PlaneSurface.cpp:176]:
```suggestion
    ActsScalar tolerance) const {
```

File: Tests/UnitTests/Core/Propagator/NavigatorTests.cpp [hunk 12]
```diff
@@ -208,6 +208,8 @@ struct PropagatorState {
     size_t debugMsgWidth = 50;
 
     const Acts::Logger& logger = Acts::getDummyLogger();
+
+    double targetTolerance = s_onSurfaceTolerance;
```
Comment (andiwand) [inline Tests/UnitTests/Core/Propagator/NavigatorTests.cpp:212]:
```suggestion
    ActsScalar targetTolerance = s_onSurfaceTolerance;
```

File: Tests/UnitTests/Core/Surfaces/SurfaceStub.hpp [hunk 13]
```diff
@@ -83,7 +83,8 @@ class SurfaceStub : public Surface {
   SurfaceIntersection intersect(const GeometryContext& /*gctx*/,
                                 const Vector3& /*position*/,
                                 const Vector3& /*direction*/,
-                                const BoundaryCheck& /*bcheck*/) const final {
+                                const BoundaryCheck& /*bcheck*/,
+                                const double /*tolerance*/) const final {
```
Comment (andiwand) [inline Tests/UnitTests/Core/Surfaces/SurfaceStub.hpp:87]:
```suggestion
                                const ActsScalar /*tolerance*/) const final {
```

**Pull request merged**